test/cpp/end2end/connection_delay_injector.cc

```diff
@@ -38,22 +38,23 @@ grpc_tcp_client_vtable* g_original_vtable = nullptr;
 grpc_core::Mutex* g_mu = nullptr;
 ConnectionAttemptInjector* g_injector ABSL_GUARDED_BY(*g_mu) = nullptr;
 
-void TcpConnectWithDelay(grpc_closure* closure, grpc_endpoint** ep,
-                         grpc_pollset_set* interested_parties,
-                         const grpc_channel_args* channel_args,
-                         const grpc_resolved_address* addr,
-                         grpc_core::Timestamp deadline) {
+int64_t TcpConnectWithDelay(grpc_closure* closure, grpc_endpoint** ep,
+                            grpc_pollset_set* interested_parties,
+                            const grpc_channel_args* channel_args,
+                            const grpc_resolved_address* addr,
+                            grpc_core::Timestamp deadline) {
   grpc_core::MutexLock lock(g_mu);
   if (g_injector == nullptr) {
     g_original_vtable->connect(closure, ep, interested_parties, channel_args,
                                addr, deadline);
-    return;
+    return 0;
   }
   g_injector->HandleConnection(closure, ep, interested_parties, channel_args,
                                addr, deadline);
+  return 0;
 }
 
-grpc_tcp_client_vtable kDelayedConnectVTable = {TcpConnectWithDelay};
+grpc_tcp_client_vtable kDelayedConnectVTable = {TcpConnectWithDelay, nullptr};
```

Comment:
In this case, we should ideally provide a cancellation function that checks whether the attempt has actually been started, and if so, calls `g_original_vtable->cancel_connect()`.  If the attempt has not actually been started, it should somehow mark the request as cancelled, so that when the request is resumed, it will not actually proceed.

A TODO is fine for now.

Reply:
Added a TODO. I'll make the change in another PR.